# 理解异步编程

- [理解异步编程](#理解异步编程)
  - [JS 单线程](#js-单线程)
  - [异步场景](#异步场景)
    - [定时器 EventLooP](#定时器-eventloop)
    - [Event Loop](#event-loop)
      - [浏览器](#浏览器)
      - [Node.js](#nodejs)
    - [网络请求](#网络请求)
    - [事件绑定](#事件绑定)
    - [Promise](#promise)
- [异步编程方法](#异步编程方法)
  - [回调函数](#回调函数)
  - [事件发布/订阅](#事件发布订阅)
    - [理解](#理解)
    - [实现](#实现)
    - [Node.js 的发布与订阅](#nodejs-的发布与订阅)
  - [Promise](#promise-1)
    - [规范](#规范)
    - [ES6 Promise API](#es6-promise-api)
      - [静态方法](#静态方法)
      - [实例方法](#实例方法)
      - [实践](#实践)
  - [Generator 生成器 //ES6 异步编程解决方案](#generator-生成器-es6-异步编程解决方案)
    - [生成器对象:](#生成器对象)
    - [next](#next)
      - [next 执行](#next-执行)
      - [next 参数](#next-参数)
      - [yield](#yield)
      - [return(param)](#returnparam)
      - [throw(param)](#throwparam)
    - [generator 函数应用](#generator-函数应用)
    - [Iterator 迭代器](#iterator-迭代器)
    - [Thunk 函数](#thunk-函数)
  - [async 函数](#async-函数)
  - [await 函数](#await-函数)

同步: 调用之后得到结果, 再进行下一个任务

异步: 调用之后先进行下一个任务, 再返回之前的结果

> Linux 常用命令
> ps //process status 查看进程状态
> top //table of processes 查看动态进程变化

## JS 单线程

JS 是单线程, 通过**浏览器内核多线程**来实现异步

> 这是因为 Javascript 这门脚本语言诞生的使命所致：JavaScript 为处理页面中用户的交互，以及操作 DOM 树、CSS 样式树来给用户呈现一份动态而丰富的交互体验和服务器逻辑的交互处理。如果 JavaScript 是多线程的方式来操作这些 UI DOM，则可能出现 UI 操作的冲突； 如果 Javascript 是多线程的话，在多线程的交互下，处于 UI 中的 DOM 节点就可能成为一个临界资源，假设存在两个线程同时操作一个 DOM，一个负责修改一个负责删除，那么这个时候就需要浏览器来裁决如何生效哪个线程的执行结果。当然我们可以通过锁来解决上面的问题。但为了避免因为引入了锁而带来更大的复杂性，Javascript 在最初就选择了单线程执行。

> 其中 Chromium 作为浏览器拥有五个进程: 浏览器进程, 渲染进程, GPU 进程, 网络进程和插件进程.

**渲染进程**包括:

- **GUI 线程** //渲染页面布局

  - 负责渲染浏览器界面 HTML 元素,当界面需要重绘(Repaint)或由于某种操作引发回流(reflow)时,该线程就会执行。在 Javascript 引擎运行脚本期间,GUI 渲染线程都是处于挂起状态的,也就是说被”冻结”了.

  - 和 JS 引擎线程互斥: 由于 JavaScript 是可操纵 DOM 的，如果在修改这些元素属性同时渲染界面（即 JavaScript 线程和 UI 线程同时运行），那么渲染线程前后获得的元素数据就可能不一致了。因此为了防止渲染出现不可预期的结果，浏览器设置 GUI 渲染线程与 JavaScript 引擎为互斥的关系，当 JavaScript 引擎执行时 GUI 线程会被挂起，GUI 更新会被保存在一个队列中等到引擎线程空闲时立即被执行。

- **JS 引擎线程** //解析&执行 JS, 与 GUI 互斥线程(因为都要操作 DOM)

  - 也可以称为 JS 内核，主要负责处理 Javascript 脚本程序，例如 V8 引擎。Javascript 引擎线程理所当然是负责解析 Javascript 脚本，运行代码。
  - JS 阻塞页面加载:由于 GUI 渲染线程与 JavaScript 执行线程是互斥的关系，当浏览器在执行 JavaScript 程序的时候，GUI 渲染线程会被保存在一个队列中，直到 JS 程序执行完成，才会接着执行。因此如果 JS 执行的时间过长，这样就会造成页面的渲染不连贯，导致页面渲染加载阻塞的感觉。

- **定时触发器线程** //setTimeout & setInterval
  - 浏览器定时计数器并不是由 JavaScript 引擎计数的, 因为 JavaScript 引擎是单线程的, 如果处于阻塞线程状态就会影响记计时的准确, 因此通过单独线程来计时并触发定时是更为合理的方案。
- **事件触发线程** //将满足触发条件的事件放入任务队列
  - 当一个事件被触发时该线程会把事件添加到待处理队列的队尾，等待 JS 引擎的处理。这些事件可以是当前执行的代码块如定时任务、也可来自浏览器内核的其他线程如鼠标点击、AJAX 异步请求等，但由于 JS 的单线程关系所有这些事件都得排队等待 JS 引擎处理。
- **异步 HTTP 请求线程** //XHR 所在线程, 比如 AJAX 请求
  - 在 XMLHttpRequest 在连接后是通过浏览器新开一个线程请求， 将检测到状态变更时，如果设置有回调函数，异步线程就产生状态变更事件放到 JavaScript 引擎的处理队列中等待处理。

## 异步场景

### 定时器 EventLooP

应用场景: 防抖, 节流, 动画, 倒计时

1. 调用 webAPIs(浏览器提供的一种能力), 比如 setTimeout
2. 定时器线程计数 2s
3. 事件触发线程将定时器事件放入任务队列
4. 主线程通过 EventLoop 遍历任务队列

### Event Loop

#### 浏览器

异步实现:

1. 宏观: 浏览器多线程
2. 微观: Event Loop //事件循环

**MacroTask** - [script, setTimeout/setInterval, setImmediate, I/O, UI rendering]

**MicroTask** - [Promise, Object.observe(监听对象变化), MutationObserver(监听 DOM 结构变化), postMessage]

#### Node.js

架构:

- node-core //Nodejs API 核心 JS 库
- 绑定 //包装和暴露底层语言
- v8(Chrome 开源 js 引擎) / libuv
  > Libuv (第三方库, Node.js 异步编程的基础, Node 底层的 io 引擎, c 语言编写事件驱动库, 异步方式返回结果给 v8 引擎)

阶段 Phase:

1. times //执行 timer 回调\*\*
2. pending callbacks //系统操作回调
3. idle, prepare //内部使用
4. poll //等待新 I/O 事件\*\*
5. check //执行 setImmediate 回调\*\*
6. close callbacks //内部使用, 执行 socket onClose

**每个阶段都有一个 Callbacks 的 FIFO 队列需要执行**. 当 EL(EventLoop) 运行到一个被指定阶段时, 该阶段的 FIFO 队列会被执行, 当队列的 Callbacks 执行完或者执行的 Callbacks 数量超过该阶段的上限时, EL 才会转入下一个阶段.

**Poll Phase**

1. 计算应该被 block 多久
2. 处理 poll 队列事件
3. 处理完在没有 callbacks 的情况下, 没有 timers 才会进入 check 阶段
   > 进入 poll 阶段，如果队列不为空或者不受限，执行队列的 callback
   >
   > > 否则，检查是否设置了 setImmediate callback, 如果是，进入 check 阶段
   >
   > > 如果否，等待 callback 加入 poll 队列，这时检查定时器是否到时间。
   >
   > > > 如果否，等待 callback 加入 poll 队列
   >
   > > > 如果是，进入 timers 阶段

**process.nextTick()**

异步 node API, 单不属于 EL 阶段, 调用此方法会暂停 EL 先进行此方法回调.

### 网络请求

### 事件绑定

### Promise

# 异步编程方法

## 回调函数

## 事件发布/订阅

### 理解

publisher --发布消息--> Event Centre <--订阅消息--> subcriber

### 实现

对象来存储发布内容

发布订阅事件 -> 数组

传统 callback 写成 subscribe

| 优点 | 松耦合(可多层嵌套), 高灵活(可多次触发事件) |
| ---- | ------------------------------------------ |
| 缺点 | 无法确保消息被触发,或者触发几次            |

在 promise 出现之前, 这是非常普遍的使用方式

### Node.js 的发布与订阅

方法默认异步 `readFile`, 同步方法为 `readFileSync`

- EventEmitter
  - 事件触发与事件监听器的封装
  - const { EventEmitter } = require('events');
  - 产生事件的对象为 events.EventEmitter 的实例
  - 继承方式使用
  - emit, on, once, addEventListener, removeEventListener

## Promise

### 规范

- A+ 术语

  - promise //一个拥有 then 方法的对象或函数,
  - thenable //一个定义了 then 方法的对象或函数
  - value //任何 js 的合法值
  - error, exception //throw 语句抛出的值
  - reject, reason //一个标示 promise 被拒绝原因的值

- then 方法
  - const promise2=promise1.then(onFulfilled, onRejected);
  - then 方法的参数
    - 两个函数参数
    - onFulfilled 在 promise 完成后被调用, onRejected 在 promise 被拒绝执行后调用
    - 只能被调用一次
  - then 方法可以调用多次, 返回值必须是 promise

### ES6 Promise API

```js
new Promise(function (resolve, reject) {
  resolve(value);
  reject(reason);
});
```

#### 静态方法

| 静态方法                        | 说明                                                                                      |
| ------------------------------- | ----------------------------------------------------------------------------------------- |
| Promise.resolve(param)          | new Promise(function (resolve, reject) {resolve(param)})                                  |
| Promise.reject(reason)          | new Promise(function (resolve, reject){reject(reason)})                                   |
| Promise.all([p1,...,pn])        | 输入一组 promise 返回一个新的 promise, 当所有 promises 都是 fulfilled, 最后才是 fulfilled |
| Promise.allSettled([p1,...,pn]) | 输入一组 promise 返回一个新的 promise, 当所有 promises 状态改变后, 结果会变成 fulfilled   |
| Promise.race([p1,...,pn])       | 输入一组 promise 返回一个新的 promise, 结果 promise 的状态跟随第一个变化的 promise 状态   |

#### 实例方法

| 实例方法                                   | 说明                                                                                       |
| ------------------------------------------ | ------------------------------------------------------------------------------------------ |
| promise.then(onFulfilled, onRejected)      | 状态改变后回调, 返回新的 promise 对象                                                      |
| promise.catch(function(reason) {})         | 同 promise.then(null, onRejected), Promise 状态为 rejected 回调                            |
| promise.finally(function(reason) {//test}) | 同 promise.then(function(){//test}, function(){//test}), 无论 promise 什么状态都会继续执行 |

**注意**:

- then/catch 返回的 promise 是新的 promise, 不是原来的 promise
- promise 对象的错误会冒泡, 直到被捕获为止, 错误会被下一个 catch 捕获, 不要忘记 catch 捕捉错误
- then 方法中使用 return
- 传递函数给 then 方法需要的函数, 不要只是传其他值
- 不要把 promise 写成嵌套

#### 实践

- 根据 Promise/A+规范实现 promise，使用 promises-aplus/promises-tests 插件验证

```js
class NPromise {
  constructor(executor) {
    this.state = "pending"; // 执行executor之前promise的状态为pending
    this.value = undefined;
    this.reason = undefined;
    this.onResolvedCallbacks = []; // 主要是调用then方法时候返回的新的promise会用到
    this.onRejectedCallbacks = [];
    let resolve = (value) => {
      if (this.state === "pending") {
        this.value = value;
        // 考虑到resolve方法可能是异步执行（后于then方法执行），所以then方法的回调会先压入到一个数组中，等到resolve方法执行的时候再遍历执行这个数组内的函数
        this.onResolvedCallbacks.forEach((fn) => fn());
        this.state = "fulfilled";
      }
    };
    let reject = (reason) => {
      if (this.state === "pending") {
        this.reason = reason;
        this.onRejectedCallbacks.forEach((fn) => fn());
        this.state = "rejected";
      }
    };
    try {
      executor(resolve, reject);
    } catch (error) {
      reject(error);
    }
  }
  then(onFulfilled, onRejected) {
    // 处理传入非函数参数的情况
    onFulfilled =
      typeof onFulfilled === "function" ? onFulfilled : (value) => value;
    onRejected =
      typeof onRejected === "function"
        ? onRejected
        : (err) => {
            throw err;
          };
    let promise2 = new NPromise((resolve, reject) => {
      // promise2是原promise执行then方法后返回的新promise
      // 这里的this指向的是原promise

      if (this.state === "fulfilled") {
        setTimeout(() => {
          // 这里使用settimeout是考虑在异步情况下使用promise2（在原promise调用then之后才能执行）
          try {
            let x = onFulfilled(this.value); // 可以看到onFulfilled回调的参数是原promise在状态决定后的value值
            resolvePromise(promise2, x, resolve, reject); // 需要一个函数根据执行回调函数的情况，判断返回的新的promise2的状态
          } catch (e) {
            reject(e);
          }
        }, 0);
      }
      if (this.state === "rejected") {
        setTimeout(() => {
          try {
            let x = onRejected(this.reason);
            resolvePromise(promise2, x, resolve, reject);
          } catch (e) {
            reject(e);
          }
        }, 0);
      }
      // 在原promise的状态还没决定的情况下，将onFulfilled和onRejected回调压入待执行函数数组中
      if (this.state === "pending") {
        this.onResolvedCallbacks.push(() => {
          setTimeout(() => {
            try {
              let x = onFulfilled(this.value);
              resolvePromise(promise2, x, resolve, reject);
            } catch (e) {
              reject(e);
            }
          }, 0);
        });
        this.onRejectedCallbacks.push(() => {
          setTimeout(() => {
            try {
              let x = onRejected(this.reason);
              resolvePromise(promise2, x, resolve, reject);
            } catch (e) {
              reject(e);
            }
          }, 0);
        });
      }
    });

    // 此方法用于判断原promise执行了then中的回调函数后的返回值x，与返回的新的promise2的关系
    function resolvePromise(promise2, x, resolve, reject) {
      // 抛出循环引用错误
      if (promise2 === x) {
        return reject(new TypeError("Chaining cycle detected for promise"));
      }
      let called; // thenable的回调只能执行一个

      if (x !== null && (typeof x === "object" || typeof x === "function")) {
        // 如果执行回调返回的结果是对象或是函数
        try {
          let then = x.then;
          if (typeof then === "function") {
            // 判断这个x是不是thenable
            // 是一个thenable，执行它的then方法来获取它的状态，根据它状态来决定promise2的状态
            then.call(
              x,
              (y) => {
                // x是fulfilled状态走这里
                if (called) return;
                called = true;
                resolvePromise(promise2, y, resolve, reject); // x的value有可能又是一个promise
              },
              (r) => {
                // x是rejected状态走这里
                if (called) return;
                called = true;
                reject(r); // （reason值不会是promise）promise2继承这个状态，并且继承这个reason值
              }
            );
          } else {
            // 不是thenable，直接resolve
            resolve(x);
          }
        } catch (e) {
          if (called) return;
          called = true;
          reject(e);
        }
      } else {
        // 如果结果是普通值，新promise的状态为fulfilled，并且value为该返回值
        resolve(x);
      }
    }
    return promise2;
  }
  // catch方法
  catch(fn) {
    return this.then(null, fn);
  }
  // Promise.resolve
  static resolve(value) {
    return new NPromise((resolve, reject) => {
      resolve(value);
    });
  }
  // Promise.reject
  static reject(reason) {
    return new NPromise((resolve, reject) => {
      reject(reason);
    });
  }
  // Promise.race
  static race(promises) {
    return new NPromise((resolve, reject) => {
      promises.forEach((promise) => promise.then(resolve, reject));
    });
  }
  // Promise.all (获取所有的promise，都执行then，把结果放到数组，一起返回)
  static all(promises) {
    let arr = [];

    function processData(index, data, length, resolve) {
      arr[index] = data;
      if (index === length) {
        resolve(arr);
      }
    }
    return new NPromise((resolve, reject) => {
      promises.forEach((promise, index, promises) => {
        promise.then((data) => {
          processData(index, data, promises.length, resolve);
        }, reject);
      });
    });
  }
}
```

## Generator 生成器 //ES6 异步编程解决方案

**声明**: 通过 Function 声明
**返回值**: 符合可迭代协议和迭代器协议的生成器对象
在执行时能暂停, 又能从暂停处继续执行
**协程/微线程 Coroutine**:

- 一个线程存在多个协程, 但同时只能执行一个
- generator 函数是协程在 ES6 的实现
- Tield 挂起 x 协程(交给其他协程), next 唤醒 x 协程

### 生成器对象:

next(param)
return(param)
throw(param)

### next

#### next 执行

有 yield 暂停, 将紧跟 yield 表达式的值作为返回的对象的 value
没有 yield, 一直执行到 return, 将 return 的值作为返回的对象的 value
没有 return, 将 undefined 作为返回的对象的 value

#### next 参数

next 方法可以带一个参数, 改参数会被当做上一个 yield 表达式的返回值

```js
function* createIterator() {
  let first = yield 1;
  let second = yield first + 2;
  yield second + 3;
}
let iterator = createIterator();

iterator.next(); //{value: 1, done:false}
iterator.next(4); //{value: 6, done:false}
iterator.next(5); //{value: 8, done:false}
iterator.next(); //{value: undefined, done:true}
```

#### yield

只能出现在 generator 函数中
用来暂停和恢复生成器函数

- `yield*`
  - 委托给其他可迭代对象
  - 作用: 复用生成器

```js
function* generator1() {
  yield 1;
  yield 2;
}

function* generator2() {
  yield 100;
  yield* generator1();
  yield 200;
}

let g2 = generator2();
g2.next(); //{value: 100, done:false}
g2.next(); //{value: 1, done:false}
g2.next(); //{value: 2, done:false}
g2.next(); //{value: 200, done:false}
g2.next(); //{value: undefined, done:true}
```

#### return(param)

给定值终结遍历器, param 可省略

```js
function* createIterator() {
  yield 1;
  yield 2;
  yield 3;
}
let iterator = createIterator();

iterator.next(); //{value: 100, done:false}
iterator.return(); //{value: undefined, done:true}
iterator.next(); //{value: undefined, done:true}
```

#### throw(param)

让生成器对象内部抛出错误

```js
function* createIterator() {
  let first = yield 1;
  let second;
  try {
    second = yield first + 2;
  } catch (e) {
    second = 6;
  }
  yield second + 3;
}
let iterator = createIterator();

iterator.next(); //{value: 1, done:false}
iterator.next(10); //{value: 12, done:false}
iterator.throw(new Error("error")); //{value: 9, done:false}
iterator.next(); //{value: undefined, done:true}
```

### generator 函数应用

```js
function* readFilesByGenerator() {
  const fs = require("fs");
  const files = [
    "/Users/jerry/Library/1.json",
    "/Users/jerry/Library/2.json",
    "/Users/jerry/Library/3.json",
  ];
  let fileStr = "";
  function readFile(filename) {
    fs.readFile(filename, function (err, data) {
      console.log(data.toString());
      f.next(data.toString());
    });
  }
  yield readFile(files[0]);
  yield readFile(files[1]);
  yield readFile(files[2]);
}

const f = readFilesByGenerator();
f.next();
```

### Iterator 迭代器

有 next 方法, 执行返回结果对象(value/done)

```js
function createIterator(items) {
  var i = 0;
  return {
    next: function () {
      var done = i >= items.length;
      var value = !done ? items[i++] : undefined;
      return {
        done: done,
        value: value,
      };
    },
  };
}

var iterator = createIterator([1, 2, 3]);

iterator.next(); // {value:1, done:false}
iterator.next(); // {value:2, done:false}
iterator.next(); // {value:3, done:false}
iterator.next(); // {value:undefined, done:true}
```

- 可迭代协议
  - [Symbol.iterator]属性
  - 内置可迭代对象, 比如 string, array, map, set etc.
- 迭代器协议
  - next 方法: done 和 value

### Thunk 函数

**求值策略**:(`sum(x+1,x+2)`)
传值调用(计算 sum 前, 先计算其中的 x+1 和 x+2 的值),
传名调用(先不计算 x+1 和 x+2, 等调用到 sum 的时候再计算)

Thunk 函数是**传名调用**的实现方式之一, 可以实现自动执行 generator 函数

```js
const fs = require("fs");
const Thunk = function (fn) {
  return function (...args) {
    return function (callback) {
      return fn.call(this, ...args, callback);
    };
  };
};
const readFileThunk = Thunk(fs.readFile); //这里的 readFileThunk是上面的 function(...args)

function run(fn) {
  var gen = fn();
  // 这里较为复杂
  function next(err, data) {
    var result = gen.next(data); //{value:function, done:false}
    if (result.done) return;
    result.value(next); //这里把 next 作为 callback 传进, 实现自动调用
  }
  next();
}

const g = function* () {
  const s1 = yield readFileThunk("/Users/kitty/testgenerator/1.json");
  // 这里会读取完此文件再进行下一个文件处理
  console.log(s1.toString());
  const s2 = yield readFileThunk("/Users/kitty/testgenerator/2.json");
  console.log(s2.toString());
  const s3 = yield readFileThunk("/Users/kitty/testgenerator/3.json");
  console.log(s3.toString());
};

run(g);
```

## async 函数

- **一个语法糖**, 使异步操作更简单
- **返回值**, 是一个 promise 对象
  - return value 是 promise resolved 的 value
  - throw value 是 promise rejected 的 reason

```js
async function test() {
  return 1;
}
const p = test();
console.log(p); // Promise{<resolved>:1}
p.then(function (data) {
  console.log(data); //1
});

async function test() {
  throw new Error("error"); //作为 promise 的 reson 返回
}
const p = test();
console.log(p); // Promise{<rejected>:1}
p.then(function (data) {
  console.log(data); //1
});
```

## await 函数

- 只能出现在 async 函数内或最外层
- 等待一个 promise 对象的值
- await 的 promise 状态为 rejected, 后续执行中断

```js
async function async1() {
  console.log("async1 start");
  await async2();
  console.log("async1 end");
}
async function async2() {
  return Promise.resolve().then((_) => {
    console.log("async2 promise");
  });
}

async1();

/* ------------------- */

async function f() {
  await Promise.reject("error"); //抛出错误后不再执行后面, 想继续执行的话在后面加上 .catch(err=>{ //异常处理操作 })即可, 或者用下面的 try...catch
  console.log(1);
  await 100;
}

// try...catch
async function f() {
  try {
    await Promise.reject("error");
  } catch (e) {
    // 处理异常
  } finally {
  }
  console.log(1);
  await 100;
}
```
